client/dashboard/me/tax-details/user-tax-form.tsx

```diff
@@ -0,0 +1,410 @@
+			value={ getValue( { item: data } ) || '' }
+		/>
+	);
+}
+
+function VatInputControl( { data, field, onChange }: UserTaxFormControlProps ) {
+	const { getValue, id, label, isDisabled } = field;
+
+	return (
+		<InputControl
+			__next40pxDefaultSize
+			disabled={ isDisabled }
+			label={ label }
+			onChange={ ( value ) => onChange( { [ id ]: value } ) }
+			value={ getValue( { item: data } ) || '' }
+		/>
+	);
+}
+
+export default function UserTaxForm() {
+	const { data: countryList } = useSuspenseQuery( countryListQuery() );
+	const countryCodes = getDataFormCountryCodes( countryList );
+
+	const [ localData, setLocalData ] = useState< Partial< UserTaxFormData > >( {} );
+	const query = useSuspenseQuery( userTaxDetailsQuery() );
+	const userTaxDetails: UserTaxDetails = query.data ?? emptyUserTaxDetails;
+	const { data: geoData } = useSuspenseQuery( geoLocationQuery() );
+	const formData = useMemo( () => {
+		const serverData = {
+			country: userTaxDetails.country ?? '',
+			id: userTaxDetails.id ?? '',
+			name: userTaxDetails.name ?? '',
+			address: userTaxDetails.address ?? '',
+		};
+		return {
+			...serverData,
+			...localData,
+		};
+	}, [
+		localData,
+		userTaxDetails?.address,
+		userTaxDetails?.country,
+		userTaxDetails?.id,
+		userTaxDetails?.name,
+	] );
+	const taxName = getTaxName( countryList, formData.country ?? geoData?.country_short ?? 'GB' );
+	const { createSuccessNotice, createErrorNotice, removeNotice } = useDispatch( noticesStore );
+	const userTaxDetailsMutation = () => {
+		return mutationOptions< UserTaxDetails, UserTaxDetailsUpdateError, UserTaxDetails >( {
+			mutationFn: updateUserTaxDetails,
+			onSuccess: ( newData: UserTaxDetails ) => {
+				queryClient.setQueryData(
+					userTaxDetailsQuery().queryKey,
+					( oldData ) => oldData && { ...oldData, ...newData }
+				);
+				removeNotice( 'vat_info_notice' );
+				createSuccessNotice(
+					sprintf(
+						/* translators: %s is the name of taxes in the country (eg: "VAT" or "GST"). */
+						__( 'Your %s details have been updated!' ),
+						taxName ?? __( 'VAT' )
+					),
+					{
+						id: 'vat_info_notice',
+						type: 'snackbar',
+					}
+				);
+			},
+			onError: ( error: Error | UserTaxDetailsUpdateError | UserTaxDetailsFetchError ) => {
+				removeNotice( 'vat_info_notice' );
+				if ( error?.message?.length > 0 ) {
+					createErrorNotice( error.message, { type: 'snackbar', id: 'vat_info_notice' } );
+					return;
+				}
+				createErrorNotice(
+					sprintf(
+						/* translators: %s is the name of taxes in the country (eg: "VAT" or "GST"). */
+						__( 'An error occurred while fetching %s details.' ),
+						taxName ?? __( 'VAT' )
+					),
+					{
+						type: 'snackbar',
+						id: 'vat_info_notice',
+					}
+				);
+			},
+		} );
+	};
+
+	const mutation = useMutation< UserTaxDetails, UserTaxDetailsUpdateError, UserTaxDetails >(
+		userTaxDetailsMutation()
+	);
+	const formatUserTaxDetails = useCallback( ( data: UserTaxDetails ) => {
```

Comment:
Just a note on `useCallback` and `useMemo`: we're trying to limit there use unless we're noticing there's an actual performance issue. Unless there's like a really obvious future proof that's worth doing (e.g. adding a `useMemo` to a `<Provider>`), we find it's better not to prematurely optimise.